### ISSUE TITLE
AzureBlobStorage Env_Auth setting forces system managed identity

### ISSUE THREAD

**Issue by niklashha** [rclone/rclone#8078]
#### What is the problem you are having with rclone?

We currently use RClone to connect Azure VMs to an Azure Blob Storage. In the context of the Azure Virtual Desktop, we would like to use the identity of the logged-in user. According to the documentation, it is possible to generate credentials via _az login_ and then use them via the Env_Auth flag. According to the documentation, however, the following sequence is used when setting the **Env_Auth** to true switch:

1. Environment Variables
2. Managed Service Identity Credentials
3. Azure CLI credentials (as used by the az tool)

Since we are operating in the Azure context and working with Entra ID, every VM in Azure requires a System Managed Identity, including our AVD machine. There is no option to enforce the credentials via az tool, it always automatically uses the System Managed Identity, which has no authorization, as this lies with the user.
Setting the **Use_msi** flag to false also has no effect. According to the documentation, there is also no way to bypass the use of the MSI.

I am not sure whether this is a bug or a desired behavior.
#### What is your rclone version (output from `rclone version`)
1.67

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)
Windows Server 22 und Windows Server 25 Preview

#### Which cloud storage system are you using? (e.g. Google Drive)
Azure Blob Storage

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)
`rclone mount`

Config File
```
[configName]
type = azureblob
account = <accountName>
env_auth = true
use_msi = false
```

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by ncw** [comment]
It looks like there is  a fix for this in the SDK - there is a bit we can set to ignore the MSI credentials.

Please give this a go with the `--azureblob-disable-instance-discovery` flag or put `disable_instance_discovery=true` in your config.

[v1.69.0-beta.8310.3a40c2100.fix-8078-azureblob-credentials](https://beta.rclone.org/branch/fix-8078-azureblob-credentials/v1.69.0-beta.8310.3a40c2100.fix-8078-azureblob-credentials/) on branch [fix-8078-azureblob-credentials](https://github.com/rclone/rclone/tree/fix-8078-azureblob-credentials) (uploaded in 15-30 mins)

Here is the help for the new feature

----

#### --azureblob-disable-instance-discovery

Ignore Microsoft Entra instance metadata

This enables Azure CLI credentials to be used instead of Managed
Service Identity Credentials.

This should be set only by applications authenticating in disconnected
clouds, or private clouds such as Azure Stack. It determines whether
the credential requests Microsoft Entra instance metadata from
https://login.microsoft.com before authenticating.

Setting this to true will skip this request and rclone will use Azure CLI
credentials instead.


Properties:

- Config:      disable_instance_discovery
- Env Var:     RCLONE_AZUREBLOB_DISABLE_INSTANCE_DISCOVERY
- Type:        bool
- Default:     false


----

I also changed the main docs slightly adding this paragraph

##### Env Auth: 2. Managed Service Identity Credentials

...

If you are operating in disconnected clouds, or private clouds such as
Azure Stack you may want to set `disable_instance_discovery = true`.
If set then rclone will ignore the Managed Service Identity
credentials, and move on to the next step Azure CLI credentials

----

Hopefully this will work!

Any corrections to the docs gratefully received - I'm not an azureblob expert so may have got terminology wrong.

**Comment by niklashha** [comment]
Hi @ncw,

thank you for your response! I tested the new feature with the beta but it seems it is not working. I'm not really sure but the behavior is the same es before. Here are some further information:
```
rclone version
rclone v1.69.0-beta.8310.3a40c2100.fix-8078-azureblob-credentials
- os/version: Microsoft Windows Server 2025 Datacenter Evaluation 24H2 (64 bit)
- os/kernel: 10.0.26100.1150 (x86_64)
- os/type: windows
- os/arch: amd64
- go/version: go1.23.1
- go/linking: static
- go/tags: cmount
```
Config file share.conf:
```
[share]
type = azureblob
account = <accountName>
env_auth = true
disable_instance_discovery = true
```

I did a `az login` and over az directly i can access the blob storage. Then I tried to mount the blob over rclone with `rclone mount share:share C:\Data\share --config C:\Data\Config\share.conf` and i received the following error message:
```
2024/09/18 06:48:11 ERROR : /: Dir.Stat error: GET https://<accountName>.blob.core.windows.net/share
--------------------------------------------------------------------------------
RESPONSE 403: 403 This request is not authorized to perform this operation using this permission.
ERROR CODE: AuthorizationPermissionMismatch
--------------------------------------------------------------------------------
﻿<?xml version="1.0" encoding="utf-8"?><Error><Code>AuthorizationPermissionMismatch</Code><Message>This request is not authorized to perform this operation using this permission.
RequestId:4f723091-f01e-0058-7f96-09cffb000000
Time:2024-09-18T06:48:11.1933248Z</Message></Error>
```
The virtual machine has a system managed identity, which has no access a the storage. My user account, which I used for `az login` has full access to the storage. I'm sure that rclone picked the system managed identity instead of the az credentials. I did also a test on a virtual machine where the system managed identity has access to storage, there i don't get this error message.

If you need further information let me know.

According to your question about terminology in Azure context this kind of identity is called System assigned Managed Identity (you wrote Managed Service Identity) see [Azure Documentation](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/overview)

**Comment by ncw** [comment]
Hmmm..

What that flag does is set this in the SDK

```
// DisableInstanceDiscovery should be set true only by applications authenticating in disconnected clouds, or
// private clouds such as Azure Stack. It determines whether the credential requests Microsoft Entra instance metadata
// from https://login.microsoft.com/ before authenticating. Setting this to true will skip this request, making
// the application responsible for ensuring the configured authority is valid and trustworthy.
DisableInstanceDiscovery bool
```

Which seems like it should do the right thing to me, but maybe I've mis-understood.

What do you think?

**Comment by bendhha** [comment]
In my opinion "DisableInstanceDiscovery" only effects the verification process to the IAM/IdP. 
So in this regard, if you want that the application verifies is on its own or depends on the IdP in this example Entra ID. 

Regarding this issue, its necessary to have the option to skip the usage of the managed identity (msi), 'cause as explained there is always a managed identity activated for the VMs. Resulting in this behaviour, that the env_auth doesn't "grap" the provided tokens vom the "az cli" and rclone uses the managed identity instead.

**Comment by ncw** [comment]
I spent quite some time looking at the way the SDK authenticates, tracing the code through. It, being Microsoft authentication, is complicated!

I decided there is no way to achieve what you want using the default credentials which are in use with `env_auth = true`.

So I made a dedicated authentication for the Azure CLI tool `az`. You will need to remove `env_auth = true` and add `use_az = true` instead.

```
[configName]
type = azureblob
account = <accountName>
use_az = true
```

Please give this a go

[v1.69.0-beta.8324.fbb6687f6.fix-8078-azureblob-credentials](https://beta.rclone.org/branch/fix-8078-azureblob-credentials/v1.69.0-beta.8324.fbb6687f6.fix-8078-azureblob-credentials/) on branch [fix-8078-azureblob-credentials](https://github.com/rclone/rclone/tree/fix-8078-azureblob-credentials) (uploaded in 15-30 mins)

Confusingly there are two Azure CLI tools (of course there are!) the other one is called `azd` and `use_az` does not support `azd`, though `env_auth` does.

**Comment by bendhha** [comment]
@ncw I've tested it and its working like a charm. 

Following testing scenarios: 
1. Set "az login" with the cli first and ensured the user has RBAC rights on the resource (Blob reader eg.), mounting the container system. 
- Accessing the mounted container is working fine without errors 

2. Mounting the container to the system (user didnt logged into az cli to azure) 
- Getting errors that are looping some time until you get the error message when to try to access the folder 
![image](https://github.com/user-attachments/assets/d547deae-fa67-4ad5-ac11-7854289bf1ae) (working as intended)

3. Mounting the container to the system. Login with "az login". Access the folder over explorer, and after folder access write "az logout" in console
- The access is still permitted - cause it seems like rclone is caching the tokens. Its not a big problem for me, but regarding security concerns the authentication tokens should be rechecked. 

@niklashha still need to confirm it too.

**Comment by ncw** [comment]
Thank you for testing

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.69
 
> * The access is still permitted - cause it seems like rclone is caching the tokens. Its not a big problem for me, but regarding security concerns the authentication tokens should be rechecked.

The tokens are intended to be cached, but they expire after 1 hour I think, so that is when logout happens.

If you find any further problems I'll re-open the issue.
